### DISCUSSION TITLE
Help Needed for Structuring Variable Products with Prisma and Next.js

### DISCUSSION THREAD

**Discussion by skyking-dev** [prisma/prisma#24896] [Q&A]
### Question

Hello everyone,

I am working on a clothing store project using Prisma and Next.js. My goal is to create a system for variable products, where each product can have different colors and sizes. So far, my database schema looks like this:

```prisma
model Product {
  id               String            @id @default(cuid())
  slug             String            @unique
  name             String
  type             ProductType
  status           ProductStatus
  visibility       ProductVisibility
  description      String
  shortDescription String
  imageUrl         String

  //Single Product
  price     Decimal?
  salePrice Decimal?

  //Variable Product
  variations ProductVariation[]

  categoryId String
  category   ProductCategory @relation(fields: [categoryId], references: [id])
  createdAt  DateTime        @default(now())
  updatedAt  DateTime        @updatedAt

  @@index([slug])
  @@map("products")
}

model ProductVariation {
  id              String                @id @default(cuid())
  productId       String
  subAttributesId ProductSubAttribute[]
  price           Decimal
  salePrice       Decimal

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  Product   Product  @relation(fields: [productId], references: [id])
}

model ProductAttribute {
  id                  String                @id @default(cuid())
  slug                String                @unique
  name                String
  createdAt           DateTime              @default(now())
  updatedAt           DateTime              @default(now())
  ProductSubAttribute ProductSubAttribute[]

  @@index([id])
  @@map("productAttributes")
}

model ProductSubAttribute {
  id                 String            @id @default(cuid())
  slug               String            @unique
  name               String
  value              String
  productAttributeId String
  type               SubAttributeTypes @default(DEFAULT)
  productAttribute   ProductAttribute  @relation(fields: [productAttributeId], references: [id])

  createdAt          DateTime          @default(now())
  updatedAt          DateTime          @default(now())
  ProductVariation   ProductVariation? @relation(fields: [productVariationId], references: [id])
  productVariationId String[]

  @@index([id])
  @@map("productSubAttributes")
}

model ProductCategory {
  id       String  @id @default(cuid())
  slug     String  @unique
  name     String
  imageUrl String?

  createdAt DateTime  @default(now())
  updatedAt DateTime  @default(now())
  Product   Product[]

  @@index([slug])
  @@map("productCategories")
}

enum SubAttributeTypes {
  COLOR
  DEFAULT
}

enum ProductType {
  SIMPLE
  VARIABLE
}

enum ProductStatus {
  PUBLISHED
  ARCHIVED
}

enum ProductVisibility {
  PUBLIC
  PRIVATE
}
```

Questions:

1- Current Structure: Is my current schema correct for handling variable products? Are there any improvements you would suggest?
2- Use of JSON: Considering that variations can have multiple attributes and sub-attributes (such as color and size), is it better to store this information in a JSON field or continue using table relations?
3- Front-end Issues: Will I encounter problems fetching and manipulating this information on the front-end with the current structure? What would be the best practice to ensure efficient performance and easy data manipulation on the front-end?

Thank you in advance for any guidance or suggestions you can offer!

Thanks!

**Comment by ludralph** [comment]
Hi @99Boost 👋 

> 1- Current Structure: Is my current schema correct for handling variable products? Are there any improvements you would suggest?

Your current schema is well-structured for handling variable products with different attributes and sub-attributes. However, there are a few improvements you could consider:

- Indexing: Ensure that you have appropriate indexes on fields that are frequently queried. You already have some indexes, but you might want to review and add more if necessary.

- Enum Usage: You are using enums for ProductType, ProductStatus, and ProductVisibility, which is good for maintaining data integrity.

- Relation Mapping: Ensure that your relations are correctly mapped and that you have considered cascading deletes if necessary.

> Use of JSON: Considering that variations can have multiple attributes and sub-attributes (such as color and size), is it better to store this information in a JSON field or continue using table relations?

Using table relations is generally better for maintaining data integrity and enabling complex queries. Prisma supports working with JSON fields, but for your use case, where you have multiple attributes and sub-attributes, table relations provide more flexibility and are easier to query.

> 3. Front-end Issues: Will I encounter problems fetching and manipulating this information on the front-end with the current structure? What would be the best practice to ensure efficient performance and easy data manipulation on the front-end?

With the current structure, you should be able to fetch and manipulate the data efficiently on the front-end. Some best practices to ensure efficient performance and easy data manipulation are:

- Efficient Queries: Use Prisma's querying capabilities to fetch only the necessary data. Avoid over-fetching by selecting only the fields you need.

- Pagination: Implement pagination for large datasets to improve performance.

- Caching: Use caching mechanisms to reduce the load on your database and improve response times.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
